addons/delivery/models/delivery_carrier.py

```diff
@@ -131,16 +155,18 @@ def install_more_provider(self):
     def _is_available_for_order(self, order):
         self.ensure_one()
         order.ensure_one()
-        if not self._match_address(order.partner_shipping_id):
+        if not self._match_address(order.partner_shipping_id) or not self._match_must_have_tags(order)\
+            or not self._match_excluded_tags(order) or not self._match_weight(order) or not self._match_volume(order):
             return False
 
         if self.delivery_type == 'base_on_rule':
             return self.rate_shipment(order).get('success')
 
         return True
 
-    def available_carriers(self, partner):
-        return self.filtered(lambda c: c._match_address(partner))
+    def available_carriers(self, partner, order):
+        return self.filtered(lambda c: c._match_address(partner) and c._match_must_have_tags(order) and c._match_excluded_tags(order)
+                             and c._match_weight(order) and c._match_volume(order))
```

Comment:
You created a method _match that does all the checks. It would be useful in the future dev